polkadot/node/tracking-allocator/src/lib.rs

```diff
@@ -0,0 +1,181 @@
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Tracking/limiting global allocator. Calculates the peak allocation between two checkpoints for
+//! the whole process. Accepts an optional limit and a failure handler which is called if the limit
+//! is overflown.
+
+use core::alloc::{GlobalAlloc, Layout};
+use std::{
+	ptr::null_mut,
+	sync::atomic::{AtomicBool, Ordering},
+};
+
+struct TrackingAllocatorData {
+	lock: AtomicBool,
+	current: isize,
+	peak: isize,
+	limit: isize,
+	failure_handler: Option<Box<dyn Fn()>>,
+}
+
+impl TrackingAllocatorData {
+	#[inline]
+	fn lock(&self) {
+		loop {
+			// Try to acquire the lock.
+			if self
+				.lock
+				.compare_exchange_weak(false, true, Ordering::Acquire, Ordering::Relaxed)
+				.is_ok()
+			{
+				break
+			}
+			// We failed to acquire the lock; wait until it's unlocked.
+			//
+			// In theory this should result in less coherency traffic as unlike `compare_exchange`
+			// it is a read-only operation, so multiple cores can execute it simultaneously
+			// without taking an exclusive lock over the cache line.
+			while self.lock.load(Ordering::Relaxed) {
+				std::hint::spin_loop();
+			}
+		}
+	}
+
+	#[inline]
+	fn unlock(&self) {
+		self.lock.store(false, Ordering::Release);
+	}
+
+	fn start_tracking(&mut self, limit: isize, failure_handler: Option<Box<dyn Fn()>>) {
+		self.lock();
+		self.current = 0;
+		self.peak = 0;
+		self.limit = limit;
+		// Cannot drop it yet, as it would trigger a deallocation
+		let old_handler = self.failure_handler.take();
+		self.failure_handler = failure_handler;
+		self.unlock();
+		core::mem::drop(old_handler);
+	}
+
+	fn end_tracking(&mut self) -> isize {
+		self.lock();
+		let peak = self.peak;
+		self.limit = 0;
+		// Cannot drop it yet, as it would trigger a deallocation
+		let old_handler = self.failure_handler.take();
+		self.unlock();
+		core::mem::drop(old_handler);
+		peak
+	}
+
+	#[inline]
+	fn track(&mut self, alloc: isize) -> bool {
+		self.lock();
+		self.current += alloc;
+		if self.current > self.peak {
+			self.peak = self.current;
+		}
+		let within_limits = self.limit == 0 || self.peak <= self.limit;
+		if within_limits {
+			self.unlock()
+		}
+		within_limits
+	}
+}
+
+static mut ALLOCATOR_DATA: TrackingAllocatorData = TrackingAllocatorData {
+	lock: AtomicBool::new(false),
+	current: 0,
+	peak: 0,
+	limit: 0,
+	failure_handler: None,
+};
+
+pub struct TrackingAllocator<A: GlobalAlloc>(pub A);
+
+impl<A: GlobalAlloc> TrackingAllocator<A> {
+	// SAFETY:
+	// * The following functions write to `static mut`. That is safe as the critical section inside
+	//   is isolated by an exclusive lock.
+
+	/// Start tracking
+	/// SAFETY: Failure handler is called with the allocator being in the locked state. Thus, no
+	/// allocations or deallocations are allowed inside the failure handler; otherwise, a
+	/// deadlock will occur.
```

Comment:
Stupid question: Why is it called in locked state? Can this be avoided? Can we get rid of this unsafe here?

Reply:
Other allocations (e.g. from other threads) might still be waiting for the lock. If we unlock it at this point, that can lead to another allocation failure and reentrancy to the failure handler, then handlers start racing for socket write. Probably not the end of the world indeed, but it's still better to guarantee that only one failure handler is executed at the same time.